index.js

```diff
@@ -0,0 +1,954 @@
+                 keys.every(function(k) { return equals(other[k], self[k]); });
+        },
+
+        //  Object#concat :: StrMap a ~> StrMap a
+        'fantasy-land/concat': function(strMap) {
+          var result = {};
+          for (var k in this) result[k] = this[k];
+          for (k in strMap) result[k] = strMap[k];
+          return result;
+        },
+
+        //  Object#map :: StrMap a ~> (a -> b) -> StrMap b
+        'fantasy-land/map': function(f) {
+          var result = {};
+          for (var k in this) result[k] = f(this[k]);
+          return result;
+        },
+
+        //  Object#reduce :: StrMap a ~> ((b, a) -> b, b) -> b
+        'fantasy-land/reduce': function(f, initial) {
+          var result = initial;
+          for (var k in this) result = f(result, this[k]);
+          return result;
+        }
+      }
+    },
+
+    Function: {
+      prototype: {
+        //  Function#toString :: Function ~> () -> String
+        toString: function() { return String(this); },
+
+        //  Function#equals :: Function ~> Function -> Boolean
+        'fantasy-land/equals': function(other) { return other === this; },
+
+        //  Function#map :: (a -> b) ~> (b -> c) -> (a -> c)
+        'fantasy-land/map': function(f) {
+          var functor = this;
+          return function(x) { return f(functor(x)); };
+        },
+
+        //  Function#promap :: (b -> c) ~> (a -> b, c -> d) -> (a -> d)
+        'fantasy-land/promap': function(before, after) {
+          var profunctor = this;
+          return function(x) { return after(profunctor(before(x))); };
+        },
+
+        //  Function#ap :: (a -> b) ~> (a -> b -> c) -> (a -> c)
+        'fantasy-land/ap': function(a2b2c) {
+          var a2b = this;
+          return function(x) { return a2b2c(x)(a2b(x)); };
+        }
+      }
+    }
+
+  };
+
+  //# toString :: Showable a => a -> String
+  //.
+  //. Returns a useful string representation of its argument. The Showable
+  //. constraint indicates that applying `String` to the argument must not
+  //. raise an exception. An example of a value for which this is not true
+  //. is `Object.create(null)`.
+  //.
+  //. Dispatches to the argument's `toString` method if appropriate.
+  //.
+  //. Where practical, `equals(eval(toString(x)), x) = true`.
+  //.
+  //. ```javascript
+  //. > toString(-0)
+  //. '-0'
+  //.
+  //. > toString(['foo', 'bar', 'baz'])
+  //. '["foo", "bar", "baz"]'
+  //.
+  //. > toString({x: 1, y: 2, z: 3})
+  //. '{"x": 1, "y": 2, "z": 3}'
+  //. ```
+  var toString = (function() {
+    //  $seen :: Array Any
+    var $seen = [];
+
+    return function(x) {
+      if ($seen.indexOf(x) >= 0) return '<Circular>';
+
+      var xType = type(x);
+      var toStringMethod =
+        has(xType, implementations) &&
+        has('toString', implementations[xType].prototype) ?
+          implementations[xType].prototype.toString :
+          x.toString;
```

Comment:
Good catch! If `x` has a `@@type` property we'll use `x.toString`, but if not we'll use our `Object#toString` method even if `x` has a custom `toString` method. We shouldn't expect authors of algebraic data types to specify `@@type`, as this is (currently) just a Sanctuary convention. I hope one day Ramda and other libraries will depend on sanctuary-type-classes so the community has a canonical repository of implementations of Fantasy Land methods for built-in types. Requiring `@@type` would make this :rainbow: future less likely.

I'm not sure exactly what you're suggesting, though. If `x` is `{}`, say, we want to use our implementation rather than `x.toString`. We could use `x.toString` if it is not identical to `Object.prototype.toString`, but then we'd run into the problems noted in sanctuary-js/sanctuary#100. For example:

```javascript
> ({}).toString === Object.prototype.toString
true
> vm.runInNewContext('({})').toString === Object.prototype.toString
false
```

This leaves us trying to infer whether the function is `Object.prototype.toString`. This is how [Ramda][1] does it:

> ```javascript
> if (typeof x.toString === 'function') {
>   var repr = x.toString();
>   if (repr !== '[object Object]') {
>     return repr;
>   }
> }
> return '{' + mapPairs(x, keys(x)).join(', ') + '}';
> ```

The `'[object Object]'` check works well in practice, but I would love to discover a less hacky solution.


[1]: https://github.com/ramda/ramda/blob/v0.22.1/src/internal/_toString.js#L38-L44